### PR TITLE
Use sh to handle SIGUSR1 and replace fork+exec with spawn

### PR DESCRIPTION
**Pull request by nbdd0121**
Idea from #250, but instead of using exit codes, this uses a pipe to communicate between smithay and sh. Comments yet to be fixed.

cc @psychon @Drakulix @vberger

### CODE REVIEW

**Comment by psychon** [comment]
I didn't properly try this (just ran the following in a terminal), but this might be using a bashism (I did not look at the diff, only the command that you wrote in #250):
```
$ ls -l /bin/sh
lrwxrwxrwx 1 root root 4 10. Dez 14:23 /bin/sh -> dash
$ sh -c 'trap "exit 1" USR1; (trap "" USR1; exec Xwayland <args>) & wait' 
sh: 1: Syntax error: ")" unexpected
```
Anyway: Using a shell like this is a nice idea. I like it.

File: src/xwayland/xserver.rs [hunk 1]
```diff
@@ -290,72 +291,79 @@ fn xwayland_ready<WM: XWindowManager>(inner: &Rc<RefCell<Inner<WM>>>) {
...
+        xwayland_args.push_str(&format!(" -listen {}", socket.as_raw_fd()));
     }
-    // setup the environment: clear everything except PATH and XDG_RUNTIME_DIR
-    for (key, _) in env::vars_os() {
+    // This command let sh to:
+    // * Set up signal handler for USR1
+    // * Launch Xwayland with USR1 ignored so Xwayland will signal us when it is ready (also redirect
+    //   Xwayland's STDOUT to STDERR so its output, if any, won't distract us)
+    // * Print "S" and exit if USR1 is received
+    command.arg("-c").arg(format!(
```
Comment (psychon) [inline src/xwayland/xserver.rs:317]: Isn't `format!` allocating memory? That's not `fork`-safe (I didn't look at the full PR, only this part)

Comment (nbdd0121) [reply src/xwayland/xserver.rs:317]: There are no forks anymore, only spawn.

Comment (psychon) [reply src/xwayland/xserver.rs:317]: Oh nice. That's actually quite neat.

**Comment by vberger** [comment]
Okay, I didn't realize that using `Command` from the std did pass non-cloexec along file descriptors to the spawned child, that is indeed quite a clean approach!

The only downside I see is that it puts us with a hard-dependency on `sh`, but that's a pretty minor thing I'd say.

Indeed, the comments and documentation needs to be adapted, as well as the `anvil` code, so that we can test this before merging.

**Review by psychon — APPROVED** [review]
This PR + #254 + the following diff gets me an anvil that can successfully show `xterm`.
```diff
diff --git a/anvil/src/main.rs b/anvil/src/main.rs
index 48ada968..99211e15 100644
--- a/anvil/src/main.rs
+++ b/anvil/src/main.rs
@@ -11,8 +11,6 @@ use std::{cell::RefCell, rc::Rc};
 
 use slog::Drain;
 use smithay::reexports::{calloop::EventLoop, wayland_server::Display};
-#[cfg(feature = "xwayland")]
-pub use smithay::xwayland::LaunchHelper;
 
 #[macro_use]
 mod shaders;
@@ -32,9 +30,6 @@ mod xwayland;
 
 use state::AnvilState;
 
-#[cfg(not(feature = "xwayland"))]
-pub type LaunchHelper = ();
-
 static POSSIBLE_BACKENDS: &[&str] = &[
     #[cfg(feature = "winit")]
     "--winit : Run anvil as a X11 or Wayland client using winit.",
@@ -43,12 +38,6 @@ static POSSIBLE_BACKENDS: &[&str] = &[
 ];
 
 fn main() {
-    // The launch helper must be forked before any threads were started
-    #[cfg(feature = "xwayland")]
-    let helper = unsafe { LaunchHelper::fork() }.unwrap();
-    #[cfg(not(feature = "xwayland"))]
-    let helper = ();
-
     // A logger facility, here we use the terminal here
     let log = slog::Logger::root(
         slog_async::Async::default(slog_term::term_full().fuse()).fuse(),
@@ -63,14 +52,14 @@ fn main() {
         #[cfg(feature = "winit")]
         Some("--winit") => {
             info!(log, "Starting anvil with winit backend");
-            if let Err(()) = winit::run_winit(display, &mut event_loop, helper, log.clone()) {
+            if let Err(()) = winit::run_winit(display, &mut event_loop, log.clone()) {
                 crit!(log, "Failed to initialize winit backend.");
             }
         }
         #[cfg(feature = "udev")]
         Some("--tty-udev") => {
             info!(log, "Starting anvil on a tty using udev");
-            if let Err(()) = udev::run_udev(display, &mut event_loop, helper, log.clone()) {
+            if let Err(()) = udev::run_udev(display, &mut event_loop, log.clone()) {
                 crit!(log, "Failed to initialize tty backend.");
             }
         }
diff --git a/anvil/src/state.rs b/anvil/src/state.rs
index 38e07e9b..0c14decb 100644
--- a/anvil/src/state.rs
+++ b/anvil/src/state.rs
@@ -32,7 +32,7 @@ use smithay::xwayland::XWayland;
 use crate::udev::MyOutput;
 #[cfg(feature = "xwayland")]
 use crate::xwayland::XWm;
-use crate::{buffer_utils::BufferUtils, shell::init_shell, LaunchHelper};
+use crate::{buffer_utils::BufferUtils, shell::init_shell};
 
 pub struct AnvilState {
     pub socket_name: String,
@@ -68,7 +68,6 @@ impl AnvilState {
         #[cfg(not(feature = "udev"))] _session: Option<()>,
         #[cfg(feature = "udev")] output_map: Option<Rc<RefCell<Vec<MyOutput>>>>,
         #[cfg(not(feature = "udev"))] _output_map: Option<()>,
-        _helper: LaunchHelper,
         log: slog::Logger,
     ) -> AnvilState {
         // init the wayland connection
@@ -174,7 +173,6 @@ impl AnvilState {
                 display.clone(),
                 &mut (),
                 log.clone(),
-                _helper,
             )
             .unwrap()
         };
diff --git a/anvil/src/udev.rs b/anvil/src/udev.rs
index 0a7eb9c3..e5a8b596 100644
--- a/anvil/src/udev.rs
+++ b/anvil/src/udev.rs
@@ -66,7 +66,6 @@ use crate::buffer_utils::BufferUtils;
 use crate::glium_drawer::{schedule_initial_render, GliumDrawer};
 use crate::shell::{MyWindowMap, Roles};
 use crate::state::AnvilState;
-use crate::LaunchHelper;
 
 #[derive(Clone)]
 pub struct SessionFd(RawFd);
@@ -94,7 +93,6 @@ type RenderSurface = FallbackSurface<
 pub fn run_udev(
     display: Rc<RefCell<Display>>,
     event_loop: &mut EventLoop<AnvilState>,
-    helper: LaunchHelper,
     log: Logger,
 ) -> Result<(), ()> {
     let name = display
@@ -131,7 +129,6 @@ pub fn run_udev(
         buffer_utils.clone(),
         Some(session),
         Some(output_map.clone()),
-        helper,
         log.clone(),
     );
 
diff --git a/anvil/src/winit.rs b/anvil/src/winit.rs
index d7cab896..eafbfcf2 100644
--- a/anvil/src/winit.rs
+++ b/anvil/src/winit.rs
@@ -20,12 +20,10 @@ use slog::Logger;
 use crate::buffer_utils::BufferUtils;
 use crate::glium_drawer::GliumDrawer;
 use crate::state::AnvilState;
-use crate::LaunchHelper;
 
 pub fn run_winit(
     display: Rc<RefCell<Display>>,
     event_loop: &mut EventLoop<AnvilState>,
-    helper: LaunchHelper,
     log: Logger,
 ) -> Result<(), ()> {
     let (renderer, mut input) = winit::init(log.clone()).map_err(|err| {
@@ -60,7 +58,6 @@ pub fn run_winit(
         buffer_utils,
         None,
         None,
-        helper,
         log.clone(),
     );
 
diff --git a/src/backend/winit.rs b/src/backend/winit.rs
index 8bd2fce8..76f606d7 100644
--- a/src/backend/winit.rs
+++ b/src/backend/winit.rs
@@ -252,6 +252,7 @@ impl CursorBackend for WinitGraphicsBackend {
     type Error = ();
 
     fn set_cursor_position(&self, x: u32, y: u32) -> ::std::result::Result<(), ()> {
+        return Ok(());
         debug!(self.logger, "Setting cursor position to {:?}", (x, y));
         self.window
             .window()
```

**Comment by psychon** [comment]
> Indeed, the comments and documentation needs to be adapted,

I'm not quite sure what the above refers to (and didn't try to look at the code/docs to see what it could refer to), but it seems to be the only remaining thing.

**Comment by nbdd0121** [comment]
Comments updated.

File: src/xwayland/xserver.rs [hunk 2]
```diff
@@ -290,72 +300,79 @@ fn xwayland_ready<WM: XWindowManager>(inner: &Rc<RefCell<Inner<WM>>>) {
     }
 }
 
-pub(crate) enum Void {}
-
 /// Exec XWayland with given sockets on given display
 ///
 /// If this returns, that means that something failed
```
Comment (vberger) [inline src/xwayland/xserver.rs:305]: This doc comment needs to be updated as well

**Review by vberger — CHANGES_REQUESTED** [review]
Thanks, it looks quite good like this! :+1: 

There is just one more comment that needs to be updated, I marked it as a review comment.

File: src/xwayland/xserver.rs (hunk 2 above)
Comment (nbdd0121) [reply src/xwayland/xserver.rs]: done

**Comment by vberger** [comment]
Okay, sorry for the delay, all is good, thank you!

**Pull request merged**